cli/src/bin/phylum.rs

```diff
@@ -180,45 +181,68 @@ async fn handle_commands() -> CommandResult {
         .await;
     }
 
-    let mut api = PhylumApi::new(
-        &mut config.auth_info,
-        &config.connection.uri,
-        timeout,
-        ignore_certs,
-    )
-    .await
-    .context("Error creating client")?;
-
-    // PhylumApi may have had to log in, updating the auth info so we should save the config
-    save_config(&config_path, &config).with_context(|| {
-        format!(
-            "Failed to save configuration to '{}'",
-            config_path.to_string_lossy()
-        )
-    })?;
+    // This is `Future<Output = PhylumApi>`. Commands that require access to the API
+    // will await on this, so that the API is not instantiated ahead of time for subcommands
+    // that don't require it.
+    let api = api_factory(&mut config, &config_path, timeout, ignore_certs);
+
+    let (subcommand, matches) = matches.subcommand().unwrap();
+    match subcommand {
+        "version" => handle_version(&app_name, ver),
+        "update" => handle_update(matches).await,
+        "parse" => handle_parse(matches),
+        "ping" => handle_ping(api.await?).await,
+        "project" => handle_project(&mut api.await?, matches).await,
+        "package" => handle_get_package(&mut api.await?, &config.request_type, matches).await,
+        "history" => handle_submission(&mut api.await?, config, matches).await,
```

Comment:
_Nothing to change here. I'm just intrigued by this_

I'm amazed that the borrow checker is okay with this. `api` holds a mutable reference to `config` until the `.await` happens. Which is just barely in time because this function consumes `config`...

If the arguments were in the opposite order, I'd guess you'd have to split it onto two lines to get it to compile.